orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,489 @@
+/*
+Copyright IBM Corp. 2022 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	ab "github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/bccsp"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"github.com/hyperledger/fabric/internal/pkg/identity"
+	"github.com/hyperledger/fabric/protoutil"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+	Close()
+}
+
+type attestationErrorAndResponse struct {
+	err  error
+	resp *orderer.BlockAttestationResponse
+}
+
+type attestationConnectionInfo struct {
+	stream *impatientAttestationStream
+	conn   *grpc.ClientConn
+}
+
+//go:generate mockery -dir . -name AttestationSource -case underscore -output mocks/
+type AttestationSource interface {
+	PullAttestations(seq uint64) []*orderer.BlockAttestation
+	Close()
+}
+
+type suspectSet map[string]struct{}
+
+func (s suspectSet) insert(entry string) {
+	s[entry] = struct{}{}
+}
+
+func (s suspectSet) remove(entry string) {
+	delete(s, entry)
+}
+
+func (s suspectSet) has(entry string) bool {
+	_, ok := s[entry]
+	return ok
+}
+
+type AttestationPuller struct {
+	Channel      string
+	Signer       identity.SignerSerializer
+	TLSCert      []byte
+	Dialer       Dialer
+	FetchTimeout time.Duration
+
+	// Internal state
+	openConnections []attestationConnectionInfo
+	endpoints       []EndpointCriteria
+	cancelStream    func()
+}
+
+func (p *AttestationPuller) seekNextEnvelope(startSeq uint64) (*common.Envelope, error) {
+	return protoutil.CreateSignedEnvelopeWithTLSBinding(
+		common.HeaderType_DELIVER_SEEK_INFO,
+		p.Channel,
+		p.Signer,
+		nextSeekInfo(startSeq),
+		int32(0),
+		uint64(0),
+		util.ComputeSHA256(p.TLSCert),
+	)
+}
+
+func (bf *AttestationPuller) closeEndpoints() {
+
+	// Close the streams and grpc connections
+	for _, oc := range bf.openConnections {
+		oc.stream.cancelFunc()
+		// TODO: Should we check the connection?
+		if oc.conn != nil {
+			oc.conn.Close()
+		}
+	}
+	bf.openConnections = nil
+}
+
+func (bf *AttestationPuller) Close() {
+	bf.closeEndpoints()
+}
+
+func (bf *AttestationPuller) pullAttestationBlock(ec EndpointCriteria, seq uint64) (*orderer.BlockAttestation, error) {
+	in, err := bf.seekNextEnvelope(seq)
```

Comment:
i think, `in` denotes the signed envelope, can we give proper variable name. `env` is common name in our code base to denote envelope.